### PR TITLE
fix: update subscription tracking

### PR DESCRIPTION
**Pull request by rebelchris**
We should only fire the event if subscription cycle actually changed, not any of the other changes.
I also decided to update our flags in this case so we have the right setting stored at all times.

AS-744 #done

### CODE REVIEW

File: src/routes/webhooks/paddle.ts [hunk 1]
```diff
@@ -203,10 +225,16 @@ export const paddle = async (fastify: FastifyInstance): Promise<void> => {
                 );
                 break;
               case EventName.SubscriptionUpdated:
-                await logPaddleAnalyticsEvent(
-                  eventData,
-                  AnalyticsEventName.ChangeBillingCycle,
-                );
+                if (!(await planChanged(eventData))) {
```
Comment (omBratteng) [inline src/routes/webhooks/paddle.ts:228]: For readability, would this be better?
```suggestion
				const didPlanChange = await planChanged(eventData);
                if (!didPlanChange) {
```

**Review by sshanzel — APPROVED** [review]
Don't have anything blocking 👀

File: src/routes/webhooks/paddle.ts (hunk 1 above)
Comment (rebelchris) [reply src/routes/webhooks/paddle.ts:228]: Agreed :D

**Review by omBratteng — APPROVED** [review]

File: src/routes/webhooks/paddle.ts [hunk 2]
```diff
@@ -93,7 +97,7 @@ const getUserId = async ({
   subscriptionId,
   userId,
 }: {
-  subscriptionId?: false | string | null;
+  subscriptionId?: false | string | null | unknown;
```
Comment (capJavert) [inline src/routes/webhooks/paddle.ts:100]: Why this? When you add unknown you are effectively saying I don't know the type so the other types are not really needed anymore.

Do we know this type?

File: src/routes/webhooks/paddle.ts [hunk 3]
```diff
@@ -114,6 +118,24 @@ const getUserId = async ({
...
+  const customData = data.data?.customData as { user_id: string };
+  const userId = await getUserId({
+    userId: customData?.user_id,
+    subscriptionId: 'subscriptionId' in data.data && data.data.subscriptionId,
+  });
+  const con = await createOrGetConnection();
+  const flags = await con.getRepository(User).findOne({
+    where: { id: userId },
+    select: ['subscriptionFlags'],
+  });
```
Comment (capJavert) [inline src/routes/webhooks/paddle.ts:131]: We double query the user, once in `getUserId` then again in the call in this function?

Comment (capJavert) [reply src/routes/webhooks/paddle.ts:131]: If we want to abstract anything maybe better that `getUserId` is just `getUser` and then we get user object either way.

File: src/routes/webhooks/paddle.ts (hunk 2 above)
Comment (rebelchris) [reply src/routes/webhooks/paddle.ts:100]: From the last addition it just became unknown type in their types.
For the others we do know and didn't need the unknown 😢

File: src/routes/webhooks/paddle.ts (hunk 3 above)
Comment (rebelchris) [reply src/routes/webhooks/paddle.ts:131]: Yeah was thinking of that as well, just meant more refactors on end type as it's a optional return.

Comment (rebelchris) [reply src/routes/webhooks/paddle.ts:131]: Also the early return only cares about returning the id

File: src/routes/webhooks/paddle.ts (hunk 2 above)
Comment (rebelchris) [reply src/routes/webhooks/paddle.ts:100]: My bad was wrong field in this type

File: src/routes/webhooks/paddle.ts (hunk 3 above)
Comment (rebelchris) [reply src/routes/webhooks/paddle.ts]: Alternative could be to have a seperate function for this case? (only 1 user now, but maybe future more)
Let me know if you prefer that

Comment (capJavert) [reply src/routes/webhooks/paddle.ts]: I would just do it inline in both places but if we use util then I would go with `getUser`. Neither blocking for this PR.

File: src/routes/webhooks/paddle.ts [hunk 4]
```diff
@@ -114,6 +118,24 @@ const getUserId = async ({
   return user.id;
 };
 
+const planChanged = async (data: SubscriptionUpdatedEvent) => {
+  const customData = data.data?.customData as { user_id: string };
+  const userId = await getUserId({
+    userId: customData?.user_id,
+    subscriptionId: data.data && data.data.id,
```
Comment (capJavert) [inline src/routes/webhooks/paddle.ts:125]:
```suggestion
    subscriptionId: data.data?.id,
```

**Pull request merged**